### ISSUE TITLE
Line markers not saving/FPS issue

### ISSUE THREAD

**Issue by UncagedCage** [Skretzo/runelite-plugins#34]
I've been working on a project of outlining the trigger border of dig/emote clue steps and I've noticed a few times now that after marking the outline I'll come back later to find that they've disappeared. Is this a known issue that's being worked on or is it related to adding too many lines in a short timeframe?
Also, after adding a new line it'll lag my game for about 3 seconds each line, which isn't a big deal but is kinda annoying. I've had this issue from the start so I don't think it's related to having too many lines marked (I'd guess I'm between 4-5k lines marked atm). Is this due to a bad interaction with another plugin or something?

**Comment by UncagedCage** [comment]
![MediumStepOutsideSlayerTower](https://github.com/Skretzo/runelite-plugins/assets/143442303/ab695cfc-ec28-4a3a-9bc5-02c653ee4a24)
Here's an example of a medium clue box I've had to remake three times now outside of the Mory slayer tower. To name a few more off the top of my head: there's also the north lumbridge cow pen dig step, digsite winch emote step, tower of life dig step, and the southern terrorbird dig step. From what I can recall, all the groups of lines that aren't saving started happening with medium clue steps, which I started doing after making a box around all the beginner and easy clue steps.

**Comment by Skretzo** [comment]
Yes, I think markers can disappear if they are created in quick succession. This would be because lines are competing when saving and end up overwriting each other. I should be able to fix this by saving less frequently, but not sure how often to save and what should trigger it.

It would be interesting for me to know if you are creating the markers [immediately as groups using the shift key](https://github.com/Skretzo/runelite-plugins/wiki/Line-Markers#How-to-use-Line-Markers) or doing it manually in the panel. The latter will most likely be more prone to issues, as I think there are several (unnecessary) save triggers in the panel.

Having 5000 lines could also be an issue, as it will most likely increase the time it takes to save significantly because the Json object it is stored as becomes quite big. I guess you could test this by exporting all your lines and creating a backup in a Pastebin, deleting all markers in the panel, and checking if the issue still persists.

It is really cool to see that you are creating outlines for clue steps. I hope you plan to contribute these outlines to the core Clue Scroll plugin as that plugin is much more suitable for displaying many lines than this plugin is (because they are static).

**Comment by UncagedCage** [comment]
Thanks for getting back to me so quickly. I've just been shift+right clicking to add lines and haven't really labeled the marker groups yet; eventually I'll probably want to list each group by their item ID to make it easier to implement in the Clue Scrolls plugin though.

With the few seconds of lag after adding each line it does somewhat often create a separate marker group, and almost always when I change the line edge in the plugin menu. It could be a useful feature to have the option to add lines to a pre-existing marker group instead of automatically creating a new group after four or five seconds like it currently seems to be.

When I add a new line it always goes into the Line Markers sidebar list; however, when I exit Runelite, it removes recently added lines/marker groups. I'm not sure if there's a timer that saves locally every so often or maybe the local file is too big and isn't saving new data or something. Temporary fix would be to export my markers before exiting but maybe a manual save button could work too. I'm pretty sure some of the lines I've added were over an hour before I logged off so it's not a super quick thing either.

**Comment by Skretzo** [comment]
Can you export all your ~5000 markers and send them to me? You can either save the export to a file and drag & drop it into a comment here on GitHub or upload it to Pastebin and provide the link. That way I can test if I can reproduce the same issues you are experiencing.

**Comment by Skretzo** [comment]
> It could be a useful feature to have the option to add lines to a pre-existing marker group instead of automatically creating a new group after four or five seconds like it currently seems to be.

Not quite what you wanted, but you can quickly merge two groups by renaming one of them to the exact name of the other group.

**Comment by UncagedCage** [comment]
[LineMarkers2309300920.txt](https://github.com/Skretzo/runelite-plugins/files/12775109/LineMarkers2309300920.txt)
Oh nice, I was worried I was going to have to manually change each line's marker group so renaming the group is going to be helpful. I exited the game, logged in again, and exported the attached text file so it should be in it's default state.

**Comment by UncagedCage** [comment]
[LineMarker2309301311.txt](https://github.com/Skretzo/runelite-plugins/files/12775454/LineMarker2309301311.txt)
I spent the last few hours compiling the marker groups that I had into one for each step and that seems to have fixed the not saving my data issue and the FPS issues. I'm not sure how many marker groups I have now so I attached a file listing the condensed version. I don't recall at what number of marker groups I started having FPS issues but I started having the save issues when I passed 400; 463 is when it stopped saving altogether.
I still had the same number of lines when condensing the marker groups started to improve FPS so it seems like the large amount of marker groups is what was causing issues. Not sure how that could be improved but good to know for future use.

**Comment by UncagedCage** [comment]
Been a little while, I realized that if I couldn't add new lines in-game after a certain point then I could potentially get around this by just importing the lines. So I went through and compiled the needed data and used a spreadsheet to make a list of all the lines I wanted. I don't have any issues importing the data and everything shows up correctly in-game; however, upon closing the game, it removes the imported lines from the plugin. I've attached the complete list of lines I import and was hoping to find out if this is a resolvable issue or just a limitation of the plugin. Thanks
[0CluePack.txt](https://github.com/user-attachments/files/16122758/0CluePack.txt)

**Comment by Dell63** [comment]
@UncagedCage Yeah, I have like 1400+ groups and have the same issues (fps/game stall upon adding new marker + not saving the new markers/groups). I'm going to try saving all the important markers as groups + exporting, then probably find where the files are and delete them and start the plugin from scratch (if I can manage this), then import the markers that are really important - because if the deletion of markers is caused by a limited number of groups, deleting most of the groups should do the trick no?
Thanks!
